packages/core/crates/core-network/src/network.rs

```diff
@@ -0,0 +1,896 @@
+            last_health: Health::UNKNOWN,
+            network_actions_api,
+            metric_network_health: SimpleGauge::new(
+                "core_gauge_network_health", "Connectivity health indicator").ok(),
+            metric_peers_by_quality: MultiGauge::new(
+                "core_mgauge_peers_by_quality",
+                "Number different peer types by quality",
+                &["type", "quality"]
+            ).ok(),
+            metric_peer_count: SimpleGauge::new("core_gauge_num_peers", "Number of all peers").ok()
+        };
+
+        instance
+    }
+
+    pub fn has(&self, peer: &PeerId) -> bool {
+        self.entries.contains_key(peer.to_string().as_str())
+    }
+
+    pub fn add(&mut self, peer: &PeerId, origin: PeerOrigin) {
+        let id = peer.to_string();
+        let now = current_timestamp();
+
+        // assumes disjoint sets
+        let has_entry = self.entries.contains_key(id.as_str());
+        let is_excluded = !has_entry && self.excluded.contains(id.as_str());
+
+        if ! is_excluded {
+            let is_ignored = if !has_entry && self.ignored.contains_key(id.as_str()) {
+                let timestamp = self.ignored.get(id.as_str()).unwrap();
+                if timestamp + (IGNORE_TIMEFRAME.as_millis() as u64) < now {
+                    self.ignored.remove(id.as_str());
+                    false
+                } else {
+                    true
+                }
+            } else {
+                false
+            };
+
+            if ! has_entry && ! is_ignored {
+                let mut entry = PeerStatus::new(peer.clone(), origin);
+                entry.is_public = self.network_actions_api.is_public(&peer);
+                self.refresh_network_status(&entry);
+
+                if let Some(_x) = self.entries.insert(peer.to_string(), entry) {
+                    // warn!("Evicting an existing record for {}, this should not happen!", &x);
+                }
+            }
+        }
+    }
+
+    pub fn remove(&mut self, peer: &PeerId) {
+        self.prune_from_network_status(&peer);
+        self.entries.remove(peer.to_string().as_str());
+        // TODO: remove from ignored and excluded as well?
+    }
+
+    pub fn update(&mut self, peer: &PeerId, ping_result: crate::types::Result) {
+        if let Some(existing) = self.entries.get(peer.to_string().as_str()) {
+            let mut entry = existing.clone();
+            entry.last_seen = if ping_result.is_err() {current_timestamp()} else {ping_result.ok().unwrap()} ;
+            entry.heartbeats_sent = entry.heartbeats_sent + 1;
+            // entry.is_public = self.network_actions_api.is_public(&peer);    // TODO: Reconsider whether this is necessary
+
+            if ping_result.is_err() {
+                entry.backoff = MAX_BACKOFF.max(entry.backoff.powf(BACKOFF_EXPONENT));
+                entry.quality = 0.0_f64.max(entry.quality - 0.1);
+
+                if entry.quality < 0.001 {
```

Comment:
This is an arbitrary value to account for floating point precision. It could be possible to use something like:
```rust
const QUALITY_STEP: f32 = 0.1;
```

Then it could be:
```suggestion
                if entry.quality < (QAULITY_STEP / 2) {
```

Would that be reasonable?

Reply:
Yep, that seems better to me